src/Cafe/OS/libs/nsyshid/Dimensions.cpp

```diff
@@ -0,0 +1,1140 @@
+			g_dimensionstoypad.GenerateRandomNumber(std::span<const uint8, 8>{buf.begin() + 4, 8}, sequence, q_result);
+			break;
+		}
+		case 0xB3: // Challenge
+		{
+			// Get the next number in the sequence based on the RNG from 0xB1 command
+			g_dimensionstoypad.GetChallengeResponse(std::span<const uint8, 8>{buf.begin() + 4, 8}, sequence, q_result);
+			break;
+		}
+		case 0xC0: // Color
+		case 0xC1: // Get Pad Color
+		case 0xC2: // Fade
+		case 0xC3: // Flash
+		case 0xC4: // Fade Random
+		case 0xC6: // Fade All
+		case 0xC7: // Flash All
+		case 0xC8: // Color All
+		{
+			// Send a blank response to acknowledge color has been sent to toypad
+			q_result = {0x55, 0x01, sequence};
+			q_result[3] = GenerateChecksum(q_result, 3);
+			break;
+		}
+		case 0xD2: // Read
+		{
+			// Read 4 pages from the figure at index (buf[4]), starting with page buf[5]
+			g_dimensionstoypad.QueryBlock(buf[4], buf[5], q_result, sequence);
+			break;
+		}
+		case 0xD3: // Write
+		{
+			// Write 4 bytes to page buf[5] to the figure at index buf[4]
+			g_dimensionstoypad.WriteBlock(buf[4], buf[5], std::span<const uint8, 4>{buf.begin() + 6, 4}, q_result, sequence);
+			break;
+		}
+		case 0xD4: // Model
+		{
+			// Get the model id of the figure at index buf[4]
+			g_dimensionstoypad.GetModel(std::span<const uint8, 8>{buf.begin() + 4, 8}, sequence, q_result);
+			break;
+		}
+		case 0xD0: // Tag List
+		case 0xE1: // PWD
+		case 0xE5: // Active
+		case 0xFF: // LEDS Query
+		{
+			// Further investigation required
+			cemuLog_log(LogType::Force, "Unimplemented LD Function: {:x}", command);
+			break;
+		}
+		default:
+		{
+			cemuLog_log(LogType::Force, "Unknown LD Function: {:x}", command);
+			break;
+		}
+		}
+
+		m_queries.push(q_result);
+	}
+
+	uint32 DimensionsUSB::LoadFigure(const std::array<uint8, 0x2D * 0x04>& buf, std::unique_ptr<FileStream> file, uint8 pad, uint8 index, bool lock)
+	{
+		if (lock)
+			std::shared_lock lock(m_dimensionsMutex);
+
+		const uint32 id = GetFigureId(buf);
+
+		DimensionsMini& figure = GetFigureByIndex(index);
+		figure.dimFile = std::move(file);
+		figure.id = id;
+		figure.pad = pad;
+		figure.index = index + 1;
+		figure.data = buf;
+		// When a figure is added to the toypad, respond to the game with the pad they were added to, their index,
+		// the direction (0x00 in byte 6 for added) and their UID
+		std::array<uint8, 32> figureChangeResponse = {0x56, 0x0b, figure.pad, 0x00, figure.index, 0x00, buf[0], buf[1], buf[2], buf[4], buf[5], buf[6], buf[7]};
+		figureChangeResponse[13] = GenerateChecksum(figureChangeResponse, 13);
+		m_figureAddedRemovedResponses.push(figureChangeResponse);
+
+		if (lock)
+			std::shared_lock unlock(m_dimensionsMutex);
+
+		return id;
+	}
+
+	bool DimensionsUSB::RemoveFigure(uint8 pad, uint8 index, bool save, bool lock)
+	{
+		DimensionsMini& figure = GetFigureByIndex(index);
+		if (figure.index == 255)
+		{
+			return false;
+		}
+
+		if (lock)
+			std::shared_lock lock(m_dimensionsMutex);
```

Comment:
This is not the proper way to use std::shared_lock. The scope of the variable named lock is the if statement. There's only one statement so the scope of that if statement ends immediately. So in this case it grabs the lock on construction and immediately releases it on destruction.
I recommend defining ``std::optional<std::lock_guard> lock;`` unconditionally instead and then constructing it in the if statement.
```suggestion
		std::optional<std::lock_guard> scopeLock;
		if(lock)
			scopeLock.emplace(m_dimensionsMutex);
```
Same goes for all the other places where the lock is acquired like this.

Also I think m_dimensionsMutex doesn't have to be a shared_mutex. It can probably be a regular std::mutex. From [cppreference.com](https://en.cppreference.com/w/cpp/thread/shared_mutex):
> Shared mutexes are especially useful when shared data can be safely read by any number of threads simultaneously, but a thread may only write the same data when no other thread is reading or writing at the same time.

Since many of the functions that acquire the shared lock write to the protected data it could only theoretically cause clashing if multiple threads called these functions simultaneously for example.

Reply:
Looks good. One more thing: DimensionsUSB::GetStatus accesses m_figureAddedRemovedResponses so it should acquire the lock too when it does so.